src/ms3/bs4_parser.py

```diff
@@ -3169,6 +3312,101 @@ def set_first_timesig(self, first_timesig: str):
+            for voice, onsets in voices.items():
+                for onset, tag_dicts in onsets.items():
+                    if onset >= mc_onset:
+                        continue
+                    for tag_dict in tag_dicts:
+                        if tag_dict["name"] != "Chord":
+                            continue
+                        from_chord_to_rest(tag_dict["tag"])
+
+    def replace_notes_after_onset(
+        self,
+        end_mc: int,
+        mc_onset: Fraction,
+        exclude_end: bool = False,
+    ):
+        staves = self.tags[end_mc]
+        for staff, voices in staves.items():
+            for voice, onsets in voices.items():
+                for onset, tag_dicts in onsets.items():
+                    if onset == mc_onset and not exclude_end:
+                        continue
+                    elif onset < mc_onset:
+                        continue
+                    for tag_dict in tag_dicts:
+                        if tag_dict["name"] != "Chord":
+                            continue
+                        from_chord_to_rest(tag_dict["tag"])
+
+    def enforce_tempo(
+        self,
+        piece_tempo_tag: Optional[bs4.Tag] = None,
+        user_tempo: Optional[float] = None,
+        user_beat_factor: Optional[Fraction] = Fraction(1 / 1),
+        user_call: Optional[bool] = True,
+    ):
+        for measure_tag in self.iter_first_measures():
+            tempo_tag = measure_tag.find("Tempo")
+            timesig_tag = measure_tag.find("TimeSig")
+            if not user_call and piece_tempo_tag is not None and tempo_tag is None:
+                # Copying active tempo tag from "parent" piece
+                timesig_tag.insert_after(piece_tempo_tag)
+                return
+            elif user_call and tempo_tag is not None:
+                # since MuseScore's reference is quarter-beats
+                relative_tempo = np.round((user_tempo / 60) * user_beat_factor, 3)
+                tempo_tag.clear()
+                _ = self.new_tag(
+                    name="tempo",
+                    value=str(relative_tempo),
+                    append_within=tempo_tag,
+                )
+                return
+            elif user_call and tempo_tag is None:
+                # since MuseScore's reference is quarter-beats
+                relative_tempo = np.round((user_tempo / 60) * user_beat_factor, 3)
+                tempo_tag = self.new_tag(name="Tempo", after=timesig_tag)
+                _ = self.new_tag(
+                    name="tempo",
+                    value=str(relative_tempo),
+                    append_within=tempo_tag,
+                )
+                return
+            elif piece_tempo_tag is None and not user_call:
+                self.logger.warning(
+                    "No active tempo was found and none was set by the user."
+                )
+                return
+
+    def decompose_repeat_tags(self, soup: bs4.BeautifulSoup):
```

Comment:
(see above) do not take soup as an argument because it is a field (`self.soup`).